src/temp_sensor.rs

```diff
@@ -0,0 +1,165 @@
+        match value {
+            TemperatureSensorClockSource::XTAL => {
+                soc_periph_temperature_sensor_clk_src_t_TEMPERATURE_SENSOR_CLK_SRC_XTAL
+            }
+            TemperatureSensorClockSource::RcFast => {
+                soc_periph_temperature_sensor_clk_src_t_TEMPERATURE_SENSOR_CLK_SRC_RC_FAST
+            }
+            TemperatureSensorClockSource::Default => {
+                soc_periph_temperature_sensor_clk_src_t_TEMPERATURE_SENSOR_CLK_SRC_DEFAULT
+            }
+        }
+    }
+}
+
+#[cfg(not(any(esp32)))]
+impl From<temperature_sensor_clk_src_t> for TemperatureSensorClockSource {
+    fn from(value: temperature_sensor_clk_src_t) -> Self {
+        match value {
+            #[allow(non_upper_case_globals)]
+            soc_periph_temperature_sensor_clk_src_t_TEMPERATURE_SENSOR_CLK_SRC_XTAL => Self::XTAL,
+            #[allow(non_upper_case_globals)]
+            soc_periph_temperature_sensor_clk_src_t_TEMPERATURE_SENSOR_CLK_SRC_RC_FAST => {
+                Self::RcFast
+            }
+            // TODO: Perhaps the default value should be mapped explicitly
+            // and all other (u32) values should cause a failure
+            _ => Self::Default,
+        }
+    }
+}
+
+// -- TemperatureSensorConfig --
+
+#[cfg(not(any(esp32)))]
+#[derive(Copy, Clone)]
+/// Rust wrapper for `temperature_sensor_config_t`
+pub struct TemperatureSensorConfig {
+    // TODO: check int size
+    pub range_min: i32,
+    pub range_max: i32,
+    pub clk_src: TemperatureSensorClockSource,
+}
+
+#[cfg(not(any(esp32)))]
+impl From<temperature_sensor_config_t> for TemperatureSensorConfig {
+    fn from(value: temperature_sensor_config_t) -> Self {
+        TemperatureSensorConfig {
+            range_min: value.range_min,
+            range_max: value.range_max,
+            clk_src: value.clk_src.into(),
+        }
+    }
+}
+
+#[cfg(not(any(esp32)))]
+impl Into<temperature_sensor_config_t> for TemperatureSensorConfig {
+    fn into(self) -> temperature_sensor_config_t {
+        temperature_sensor_config_t {
+            clk_src: self.clk_src.into(),
+            range_max: self.range_max,
+            range_min: self.range_min,
+        }
+    }
+}
+
+#[cfg(not(any(esp32)))]
+impl Default for TemperatureSensorConfig {
+    fn default() -> Self {
+        TemperatureSensorConfig {
+            range_min: -10,
+            range_max: 80,
+            clk_src: TemperatureSensorClockSource::Default,
+        }
+    }
+}
+
+// -- TemperatureSensorDriver --
+
+#[cfg(not(any(esp32)))]
+pub struct TemperatureSensorDriver {
+    ptr: temperature_sensor_handle_t,
+    // To track current state (to avoid double enable/disable)
+    enabled: bool,
+}
+
+#[cfg(not(any(esp32)))]
+impl TemperatureSensorDriver {
+    pub fn new(config: TemperatureSensorConfig) -> Self {
+        let mut sensor = core::ptr::null_mut();
+        unsafe {
+            temperature_sensor_install(&config.into(), &mut sensor);
```

Comment:
don,'t eat the errors you get back from the underlying C api calls, every Error should be propagated. You can use a `Result<(),EspError> on the return's of all the method calls where its needed

Reply:
for error converting we have the esp! macro

Reply:
Done in 2b0694d067e75c1ffbe118c3a890e6b6801fb5c8. The only thing I'm still uncertain about is how to handle errors that occur in drop, since drop has to return (). Perhaps you have a good idea.

Reply:
yeah you don't have many chances in a drop. We either eat the error, unwrap it or panic if we get one. The underlying driver probably still sends something out over Error println. Out of my head in most cases we eat the error in the othere drivers here.

Reply:
No, we panic